### PR TITLE
paginate eth logs

### PR DESCRIPTION
**Pull request by ayrat555**
resolves https://github.com/poanetwork/blockscout/issues/2353

### Changelog

- paginate eth logs

### CODE REVIEW

File: apps/explorer/lib/explorer/eth_rpc.ex [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+        {:error, "Balance not found"}
+    end
+  end
+
+  def eth_get_logs(filter_options) do
+    with {:ok, address_or_topic_params} <- address_or_topic_params(filter_options),
+         {:ok, from_block_param, to_block_param} <- logs_blocks_filter(filter_options),
+         {:ok, from_block} <- cast_block(from_block_param),
+         {:ok, to_block} <- cast_block(to_block_param),
+         {:ok, paging_options} <- paging_options(filter_options["paging_options"]) do
```
Comment (pasqu4le) [inline apps/explorer/lib/explorer/eth_rpc.ex]: Just for consistency with the other functions called above, I'd have the `paging_options` function take the whole `filter_options` as well.

File: apps/explorer/lib/explorer/etherscan/logs.ex [hunk 2]
```diff
@@ -231,4 +241,19 @@ defmodule Explorer.Etherscan.Logs do
   end
 
   defp where_multiple_topics_match(query, _, _, _), do: query
+
+  defp page_logs(query, %{block_number: nil, transaction_index: nil, log_index: nil}) do
+    query
+    |> order_by([log], asc: log.index)
```
Comment (pasqu4le) [inline apps/explorer/lib/explorer/etherscan/logs.ex]: Given that the logs are ordered this way in both cases, couldn't this ordering be put in `query_with_block_transaction_data`, that already has an `order_by`, instead?

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,6 +1,7 @@
 ## Current
 
 ### Features
+- [#2366)(https://github.com/poanetwork/blockscout/pull/2366) - paginate eth logs
```
Comment (saneery) [inline CHANGELOG.md]: typo with the closing bracket `[#2366)`

**Comment by ayrat555** [comment]
@vbaranov 
1. I added an example with paging options.
2. Did you include `fromBlock`, `toBlock` params?
try `curl -v -X POST --data '{"id": 0, "jsonrpc": "2.0", "method": "eth_getLogs", "params": [{"address": "0xc78Be425090Dbd437532594D12267C5934Cc6c6f", "fromBlock": "earliest", "toBlock": "latest", "topics": ["0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef"]}]}' "http://localhost:4000/api/eth_rpc"`
response contains exactly 1_000 items
3. I updated a note
4. I changed param nates to macth json rpc spec

**Comment by vbaranov** [comment]
@ayrat555 

> Did you include fromBlock, toBlock params?

Yes, that is what the issue.

1 - 4 ✔️. One more thing in order to improve pagination UX: is it possible (if no performance concerns) to reorder response in that way, that all three pagination's params would be together? Therefore, the user can copy them in one click. Because currently, a user should copy every parameter to collect them all to make a request for the next page.

**Comment by ayrat555** [comment]
@vbaranov  it's possible. but do you think users will use pagination params not programmatically but by hand?

**Comment by vbaranov** [comment]
> @vbaranov it's possible. but do you think users will use pagination params not programmatically but by hand?

Got your point. It makes sense, that programmatic control is prevalent for API. If API users will have this feature request, let's implement it. But for now, let's leave it as is.

**Pull request merged**